Feature: Public portal customizations

MatiasArriola opened the pull request:
### :pushpin: References

-   **Issue:** Closes https://app.clickup.com/t/8696vv89u #8696vv89u

### :memo: Implementation

- Implemented new visualizations plugin
- Updated DHIS and EyeSeeTea dependencies
- Added Organization Unit Filters
- Moved settings to separate page (admis only)
- Display DHIS header for admins only
- Added Header and Footer customizations
- Conditionally use the `Plugin` component from `app-runtime` depending on installed app versions

TODO: 
- [x] Review and fix bundle size increase
- [x] Fix horizontal loaders not hiding after postRobot errors `Error: Window did not load after 15000ms`
- [x] Evaluate removing legacy plugin (v40+ support)
- [x] Review visualization-specific issues (year by year)

MatiasArriola committed:
feat: Add new iframe plugin implementation

Keep js plugins as LegacyPlugins
Load Legacy Plugins conditionally
Load Visualizations independently (new repo)
Move eventReport and eventChart to eventVisualization

MatiasArriola committed:
Fix report for some pivot tables
Check version to use legacy only for dhis < 39
Refactor and small Type changes

MatiasArriola committed:
docs: Update "About Plugins" readme section

MatiasArriola committed:
refactor: if/elseif/else code style, favor use of Maybe and Id, i18n error message

eperedo committed:
Merge pull request #15 from EyeSeeTea/feature/new-visualization-plugin

Feature: new visualization plugin

MatiasArriola committed:
feat: Add Org Units filter
upgrade dependencies

MatiasArriola committed:
feat: Add custom header and footer. Hide dhis HeaderBar for non-admin users.

MatiasArriola committed:
feat: Move settings to its own page, just for Admins

MatiasArriola committed:
refactor: simplify PluginVisualizationD2Repository filters, avoid code duplication

MatiasArriola committed:
refactor: get only displayName for OrgUnit. Improve variable naming and typings

eperedo committed:
Merge pull request #16 from EyeSeeTea/feat/org-unit-selector

feat: Add Org Units filter

MatiasArriola committed:
fix: Moved config options to readme instead of having commented code

MatiasArriola committed:
fix: check for admin in SaveSettingsUseCase
refactor: create a separate new UseCase for default settings initialization and move default settings fetch to its own repository

MatiasArriola committed:
fix: avoid i18n namespace error when using ":" in translation key

MatiasArriola committed:
Merge branch 'feat/ntd-customizations' into feat/header-footer-settings-changes

MatiasArriola committed:
fix: i18n duplicated keys

MatiasArriola committed:
refactor: extract function getSettingsOrInitialize
Convert unnecesary flatMap to map

eperedo committed:
Merge pull request #17 from EyeSeeTea/feat/header-footer-settings-changes

feat: Custom header and footer. Move settings to its own page

adrianq commented:
@MatiasArriola @Ramon-Jimenez The development branch was missing in this repo. I have restored it and changed the target branch of this pull request so we can keep our typical workflow

MatiasArriola committed:
Merge branch 'development' into feat/ntd-customizations

MatiasArriola committed:
fix: update i18n import paths to use utils/i18n

Updated to be consistent with changes coming from #24

MatiasArriola committed:
Merge branch 'development' into feat/ntd-customizations

MatiasArriola committed:
fix: fallback for userCredentials object deprecated in v42

MatiasArriola committed:
chore: update d2-api

MatiasArriola committed:
feat: remove redundant ProgressBar and improve styles for each Visualization

MatiasArriola committed:
feat: use new plugin for PIVOT_TABLEs, support virtualization on export

MatiasArriola committed:
chore: prettify

MatiasArriola committed:
feat: drop support for versions < 239

Only keep eventchart and eventreport legacy JS plugins, since they don't support newer iframe plugins.

eperedo commented on src/data/repositories/OrgUnitD2Repository.ts:
```diff
@@ -0,0 +1,48 @@
+
+const orgUnitFields = { id: true, displayName: true, path: true, level: true } as const;
+
+type D2OrgUnit = MetadataPick<{
```
Not really important, but for classes we put types and const like `orgUnitFields` and `D2OrgUnit` at the end of the file so the first thing you can check is the class implementation. I know this is a subjective thing so no need to change it.

eperedo commented on src/data/repositories/OrgUnitD2Repository.ts:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    public getOrgUnitsByIds(ids: Id[]): FutureData<OrgUnit[]> {
+        return apiToFuture(
```
We should consider doing this request in chunks to avoid the classic `414 URI Too Long`. If for our current use case is not a problem that's ok.

MatiasArriola replied:
It won't affect us in this use case because this method is only used to retrieve Org Units names that the user selected in the OrgUnitsSelector, since the control does not return the names. The user should pick a rather extensive list to hit the URI too long issue. Still a good point, thanks.

eperedo commented on src/domain/entities/OrgUnit.ts:
```diff
@@ -0,0 +1,21 @@
+    level: number;
+}
+
+const pathSeparator = "/";
```
In other projects we're using `path` as an array in the domain because the logic around the `pathSeparator` is a DHIS thing that we should not be in the domain (if we use a different implementation that uses underscore as separator we would need to change our domain code).

It's not a big deal here since this does not have a direct impact on the visualization logic, but you can make the change if we have the time.

eperedo commented on src/webapp/components/org-units-filter/OrgUnitPicker.tsx:
```diff
@@ -0,0 +1,37 @@
+                filterByProgram: false,
+                selectAll: false,
+            }}
+            onChildrenLoaded={{}}
```
Why do we need an empty object for this prop? Can we just ignore the prop or is this required in ui-components?

MatiasArriola replied:
I recall adding it because I was hitting an undefined reference in the OrgUnitsSelector, but I tried removing it now and it works. So we are better removing it.

eperedo commented on src/webapp/hooks/useOrgUnitRoots.ts:
```diff
@@ -0,0 +1,30 @@
+        }
+        const cancel = fetchOrgUnitRoots();
+
+        return () => cancel();
```
Not wrong, but I think we can directly call `return compositionRoot.orgUnits.getRoots.execute()` inside the useEffect and the `Future` implementation will handle the cancellation, right?

eperedo reviewed (changes requested):
[code-review] Thanks @MatiasArriola, just minor things. Let me know if you have any questions.

MatiasArriola committed:
refactor: change OrgUnitPath to Id[]

Remove unnecesary wrapper in useOrgUnitRoots
OrgUnitD2Repo: move typings after implementation
Remove unused OrgUnitsSelector onChildrenLoaded

MatiasArriola committed:
feat: add publicPortalMode flag

The DHIS2 header will be hidden only if this flag is true and the user is not an admin

adrianq committed:
bump version

MatiasArriola committed:
feat: add support for app-runtime Plugin component

Compatibility for newer app versions, related: https://github.com/dhis2/dashboard-app/releases/tag/v101.0.0
For older versions, keep using iframe + postRobot

MatiasArriola committed:
fix: give precedence to installed apps
fix: support absolute urls in `useDhis2Url`

If a different version of a bundled core app is installed, resolve to that version only. Previously AppD2Repo.get() returned multiple copies of the same app (bundled + installed).
This fixes versionsing / loading plugins conflicts.

MatiasArriola committed:
fix: load ExtJS dependency globally for legacy plugins

MatiasArriola committed:
feat: line listing support for export
fix: update LegacyReportType to reflect that reportTablePlugin chartPlugin and mapPlugin legacy support was removed.
fix: useDhis2Url for dev environment and cross-origin errors

MatiasArriola committed:
chore: rename ext-all.js to ext-all.min.js to avoid prettier

MatiasArriola committed:
fix: empty brackets in line_list visualizations

Changed the properties passed to the plugin to match dashboard-app behavior

adrianq committed:
bump version
